### PR TITLE
Storybook 7 upgrade

### PR DESCRIPTION
**Pull request by jasonbasuil**
Initially referenced [Upgrade Storybook 6.5 and Node >= 16 < 17](https://github.com/user-interviews/ui-design-system/pull/898) to get setup with upgrading Node.

This PR goes the extra step further to get us on Storybook 7 (jumping from `@storybook v6.2.1`) I think this gets us most of the way there. I mostly followed the [Storybook migration docs ](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#migration) and tried to reduce the number of deprecation warnings in the console output. 

Summary of what comes with [Storybook 7](https://storybook.js.org/blog/storybook-7-0/)

I've been able to stand this up locally as well as locally link with 'rails-server' with no major issues so far. 

Most recent Chromatic build can be found here: https://62d040e741710e4f085e0647-yoxsmdmwaj.chromatic.com/?path=/docs/intro--docs

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,20 +8,21 @@
     "react-select": "^5.0.0",
     "react-toggle": "4.1.1",
     "react-transition-group": "^4.3.0",
-    "uuid": "^7.0.2"
+    "uuid": "^7.0.2",
+    "webpack": "^5.80.0"
```
Comment (rroppolo) [inline package.json:12]: I think webpack should be a devDependency?

File: package.json [hunk 2]
```diff
@@ -133,9 +138,8 @@
     "scss"
   ],
   "engines": {
-    "node": "^ 14.17.4",
-    "npm": "^ 7.20.3",
-    "yarn": "^ 1.22.5"
+    "node": ">= 16 < 17",
```
Comment (rroppolo) [inline package.json:141]: I think this might need to be at least 16.10 given some of the other dependencies

File: package.json [hunk 3]
```diff
@@ -146,14 +150,14 @@
     "url": "git+https://github.com/user-interviews/ui-design-system.git"
   },
   "author": {
-    "name": "Dave Ferris",
-    "email": "dave@userinterviews.com"
+    "name": "Jason Basuil",
+    "email": "jason@userinterviews.com"
```
Comment (rroppolo) [inline package.json:154]: 🎉 😄

File: stories/Intro.stories.mdx [hunk 4]
```diff
@@ -16,38 +15,38 @@ import GithubLogo from '../public/github-logo.svg';
   <a
     type="button"
     style={{
-      background: colors.UX_NEUTRAL_300,
+      background: '#E6E2E0',
```
Comment (rroppolo) [inline stories/Intro.stories.mdx:18]: What happened here? Are these colors no longer accessible in mdx?

**Review by rroppolo — COMMENTED** [review]
Sorry I never got back to you last week! This is looking good though -- the only thing that jumped out on first glance was webpack in the main dependencies.

File: package.json (hunk 2 above)
Comment (jasonbasuil) [reply package.json:141]: @jeffbax are we wanting to keep this loose requirement of the node engine?  I noticed that at least we're setting `16.20.0` on the [github workflows ](https://github.com/user-interviews/ui-design-system/pull/911/files#diff-5db7bb8117a68fd50e5b5e0cbdca28ab4e360e4dd0aeba88911f9404a719700b)

Maybe `"node": ">= 16.20.0 < 17"` ?

Comment (jasonbasuil) [reply package.json:141]: @rroppolo what were some of the dependencies you spotted that required `>= 16.10` ?

Comment (rroppolo) [reply package.json:142]: Looks like it was:
```
error @jest/transform@29.5.0: The engine "node" is incompatible with this module. Expected version "^14.15.0 || ^16.10.0 || >=18.0.0". Got "16.0.0"
```

**Comment by jasonbasuil** [comment]
So it seems like [snapshot testing is failing](https://github.com/user-interviews/ui-design-system/actions/runs/4878213979/jobs/8703651491?pr=911#step:6:18) because `isBatchingLegacy` might be a deprecated function in `react-test-renderer`. 

I was pointed in the direction of upgrading some of our [react dependencies](https://copyprogramming.com/howto/react-native-web-error-rnw-blogpost-bundle-js-1414-uncaught-typeerror-cannot-read-properties-of-undefined-reading-isbatchinglegacy) which I tried and does resolve the issue, but probably is not the solution since we want to keep RS/DS dependencies in sync.  

Just flagging for now as I keep digging into it. @jeffbax @rroppolo

**Comment by jeffbax** [comment]
> So it seems like [snapshot testing is failing](https://github.com/user-interviews/ui-design-system/actions/runs/4878213979/jobs/8703651491?pr=911#step:6:18) because `isBatchingLegacy` might be a deprecated function in `react-test-renderer`.
> 
> I was pointed in the direction of upgrading some of our [react dependencies](https://copyprogramming.com/howto/react-native-web-error-rnw-blogpost-bundle-js-1414-uncaught-typeerror-cannot-read-properties-of-undefined-reading-isbatchinglegacy) which I tried and does resolve the issue, but probably is not the solution since we want to keep RS/DS dependencies in sync.
> 
> Just flagging for now as I keep digging into it. @jeffbax @rroppolo

Shouldn't a testing library be a dev dependency, and not really impact rails server?

**Comment by jasonbasuil** [comment]
> > So it seems like [snapshot testing is failing](https://github.com/user-interviews/ui-design-system/actions/runs/4878213979/jobs/8703651491?pr=911#step:6:18) because `isBatchingLegacy` might be a deprecated function in `react-test-renderer`.
> > I was pointed in the direction of upgrading some of our [react dependencies](https://copyprogramming.com/howto/react-native-web-error-rnw-blogpost-bundle-js-1414-uncaught-typeerror-cannot-read-properties-of-undefined-reading-isbatchinglegacy) which I tried and does resolve the issue, but probably is not the solution since we want to keep RS/DS dependencies in sync.
> > Just flagging for now as I keep digging into it. @jeffbax @rroppolo
> 
> Shouldn't a testing library be a dev dependency, and not really impact rails server?

Would it be fine in that case to change the dev dependencies for react but then keep the react peer dependencies the same as RS?

File: package.json [hunk 5]
```diff
@@ -71,17 +71,20 @@
...
+    "@storybook/addon-docs": "7.0.8",
+    "@storybook/addon-jest": "7.0.8",
+    "@storybook/addon-knobs": "7.0.2",
+    "@storybook/addon-links": "7.0.8",
+    "@storybook/addon-mdx-gfm": "7.0.8",
+    "@storybook/addon-storyshots": "7.0.8",
+    "@storybook/addon-storysource": "7.0.8",
+    "@storybook/addons": "7.0.8",
+    "@storybook/react": "7.0.8",
+    "@storybook/react-webpack5": "7.0.8",
```
Comment (jasonbasuil) [inline package.json:86]: upgrading to storybook@latest resolves the following critical dependency:

Screenshot 2023-05-04 at 1 19 53 PM

Fix seems to have been released just yesterday: https://github.com/storybookjs/storybook/issues/21316#issuecomment-1532765278

File: .storybook/preview.js [hunk 6]
```diff
@@ -1,12 +1,11 @@
...
-      order: ['Foundations', 'Components'],
+const preview = {
+  parameters: {
+    options: {
+      storySort: {
+        order: ['Foundations', 'Components'],
+      },
     },
-  },
+  }
```
Comment (rsaris) [inline .storybook/preview.js:8]: 💅 Do we not have the same linters as in `rails-server` that would require a trailing comma here?

**Review by rsaris — COMMENTED** [review]
This is huge! Thanks for doing all the leg work here, excited to continue the push to latest node.

Left some very unimportant comments but will leave it to the other reviewers to approve as I don't know this project well.

File: .storybook/main.js [hunk 7]
```diff
@@ -1,20 +1,37 @@
...
+    '@storybook/addon-controls',
+    '@storybook/addon-links',
+    '@storybook/addon-knobs',
+    '@storybook/addon-storysource',
+    '@storybook/addon-mdx-gfm'
   ],
+  docs: {
+    autodocs: true,
+    defaultName: 'Docs',
+  },
```
Comment (jasonbasuil) [inline .storybook/main.js:23]: Auto docs:

https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#autodocs-changes

File: src/Alert/Alert.mdx [hunk 8]
```diff
@@ -11,9 +12,7 @@ import Alert from './Alert'
   should be used cautiously.
 </h4>
 
-<Canvas>
-  <Story id="components-alert--success"/>
-</Canvas>
+<Canvas of={ComponentStories.Success} />
```
Comment (jasonbasuil) [inline src/Alert/Alert.mdx:15]: Doc blocks changed their api, no longer referencing stories by `id`.

https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#story-block

File: src/Alert/Alert.mdx [hunk 9]
```diff
@@ -25,53 +24,41 @@ import Alert from './Alert'
 
 ## Props
 
-<ArgsTable of={Alert} />
+<ArgTypes of={Alert} />
```
Comment (jasonbasuil) [inline src/Alert/Alert.mdx:28]: `ArgsTable` deprecated

https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#argstable-block

File: .storybook/main.js [hunk 10]
```diff
@@ -1,20 +1,40 @@
...
     '../stories/Intro.stories.mdx',
     '../src/**/*.stories.@(js|mdx)',
     '../src/**/*.stories.js[x]',
     '../stories/**/*.stories.mdx',
     '../stories/**/*.stories.js[x]'
-  ]
+  ],
+  features: {
+    storyStoreV7: false,
+  },
```
Comment (jasonbasuil) [inline .storybook/main.js:33]: `storyStoreV7` was an experimental optimization introduced in SB6.4 which allows for code splitting for faster build times. It's enabled by default in SB7, but I noticed that there is some lag loading fonts on all the stories (you'd have to wait maybe 15 seconds and kind of click around to different stories) and then the fonts would load properly. This also lead to some unexpected visual diffs in Chromatic. Still trying to figure out if there's a way around this.

https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#story-store-v7

Comment (jasonbasuil) [reply .storybook/main.js:33]: Two Chromatic builds with and without `storyStoreV7` (ignore the green diffs, that's just the padding that got removed broadly from all stories). Also, sorry if it's hard to see, but just showing the difference in fonts. This is just one example but font issues were seen in a lot of stories when built in Chromatic.

`storyStoreV7: false` 

Screenshot 2023-05-08 at 2 20 04 PM

`storyStoreV7: true` (default in SB7. you can kinda see we have the wrong font loaded...some serif font)

Screenshot 2023-05-08 at 2 20 51 PM

Comment (jeffbax) [reply .storybook/main.js:33]: are build times that long that we really care?

File: .storybook/preview.js [hunk 11]
```diff
@@ -1,12 +1,11 @@
-import React from 'react';
-import { addDecorator } from "@storybook/react";
-
-addDecorator(story => <div style={{ padding: '1rem' }}>{story()}</div>);
-
```
Comment (jasonbasuil) [inline .storybook/preview.js:5]: All the padding diffs in Chromatic are due to this. `addDecorator` api deprecated. https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#removed-global-client-apis

Example below:

Screenshot 2023-05-08 at 2 20 51 PM

File: .storybook/main.js (hunk 10 above)
Comment (jasonbasuil) [reply .storybook/main.js:33]: I didn't really care. Maybe it's like 20-30 seconds longer for initial build time so probably not.

**Comment by jeffbax** [comment]
> > > So it seems like [snapshot testing is failing](https://github.com/user-interviews/ui-design-system/actions/runs/4878213979/jobs/8703651491?pr=911#step:6:18) because `isBatchingLegacy` might be a deprecated function in `react-test-renderer`.
> > > I was pointed in the direction of upgrading some of our [react dependencies](https://copyprogramming.com/howto/react-native-web-error-rnw-blogpost-bundle-js-1414-uncaught-typeerror-cannot-read-properties-of-undefined-reading-isbatchinglegacy) which I tried and does resolve the issue, but probably is not the solution since we want to keep RS/DS dependencies in sync.
> > > Just flagging for now as I keep digging into it. @jeffbax @rroppolo
> > 
> > 
> > Shouldn't a testing library be a dev dependency, and not really impact rails server?
> 
> Would it be fine in that case to change the dev dependencies for react but then keep the react peer dependencies the same as RS?

I think with peer deps you want to keep things as loose as possible. Particularly building a library, you want to restrict the apps using it as little as possible

File: .storybook/preview.js (hunk 6 above)
Comment (jeffbax) [reply .storybook/preview.js:8]: probably worth adding for a 2.0 :)

File: src/LoadingOverlay/LoadingOverlay.mdx [hunk 12]
```diff
@@ -9,40 +10,17 @@ import LoadingOverlay from './LoadingOverlay';
...
 
-<Canvas>
-  <Story id="components-loadingoverlay--default" />
-</Canvas>
+<Canvas of={ComponentStories.Default} />
 
 ### When to use
-- Reason 1
-- Reason 2
+- Long loading times: If your application takes a long time to load data or process requests, using a LoadingOverlay component can help improve user experience by providing a visual indication that the application is still functioning and hasn't frozen.
```
Comment (jeffbax) [inline src/LoadingOverlay/LoadingOverlay.mdx:16]: good eye on these

**Comment by jeffbax** [comment]
@jasonbasuil is there a live preview of this build or would we just have to run it locally -- not sure if you want others to look at anything on that front or if its more of a "this all breaks or it doesn't"

**Comment by jasonbasuil** [comment]
> @jasonbasuil is there a live preview of this build or would we just have to run it locally -- not sure if you want others to look at anything on that front or if its more of a "this all breaks or it doesn't"

@jeffbax Most recent Chromatic build can be found here: https://62d040e741710e4f085e0647-zpbcxjbsot.chromatic.com/?path=/docs/intro--docs

Yeah mostly just concerned with "does it build?" and "do any of the stories / docs break?" (which I've checked every story page and fixed the issues that was causing that story / doc page to break and it seems fine currently.)

File: src/Accordion/Accordion.stories.jsx [hunk 13]
```diff
@@ -1,5 +1,4 @@
 import React from 'react';
-import { boolean, text, withKnobs } from '@storybook/addon-knobs';
```
Comment (kyleshike) [inline src/Accordion/Accordion.stories.jsx:2]: Mainly just curious: is there any situation in which we'd use knobs over controls?  Still learning about Storybook here

**Review by kyleshike — COMMENTED** [review]
This looks great.  

One small thing I noticed when doing a QA run locally is the knob count in the add ons tab doesn't seem to update correctly, not sure if this is a configuration issue or a bug with the add on?

More broadly, are knobs something we'll be moving away from in favor of controls?


https://github.com/user-interviews/ui-design-system/assets/3161151/35c198ca-4323-4e0a-b18f-7bbd5ff95970

File: src/Accordion/Accordion.stories.jsx (hunk 13 above)
Comment (jasonbasuil) [reply src/Accordion/Accordion.stories.jsx:2]: Knobs was [deprecated in SB6.3](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated-addon-knobs:~:text=6.3%20deprecations-,Deprecated%20addon%2Dknobs,-We%20are%20replacing) . I think Controls seem like they're easier to generate what used to be "knobs" in that 1) you won't need to import every knob (e.g. `text`, `boolean`, `select`, etc.) into a `Component.stories.jsx` file and 2) SB7 with Controls will generate what interactions you want to control without writing that extra code. 

I was able to make the switch to Controls in the `Accordion.stories.jsx` file with the `Default.args` obj at the bottom:

```
export const Default = args => <Accordion>
    <AccordionItem>
      <AccordionToggle eventKey="0" leadingIcon={faCreditCard} {...args} />
      <AccordionCollapse eventKey="0">
        <ul>
          <li>Item 1</li>
          <li>Item 2</li>
          <li>Item 3</li>
        </ul>
      </AccordionCollapse>
    </AccordionItem>
    <AccordionItem>
      <AccordionToggle eventKey="1" leadingIcon={faCreditCard} {...args} />
      <AccordionCollapse eventKey="1">
        <ul>
          <li>Item 1</li>
          <li>Item 2</li>
          <li>Item 3</li>
        </ul>
      </AccordionCollapse>
    </AccordionItem>
  </Accordion>;
Default.args = {
  chevronLeft: false,
  disabled: false,
  helperText: 'helper text',
  title: 'Accordion Toggle'
};
```

**Comment by jasonbasuil** [comment]
> This looks great.
> 
> One small thing I noticed when doing a QA run locally is the knob count in the add ons tab doesn't seem to update correctly, not sure if this is a configuration issue or a bug with the add on?
> 
> More broadly, are knobs something we'll be moving away from in favor of controls?
> 
>  knob.count.in.tab.mov

Yeah we'll be moving away from Knobs due to the 6.3 deprecation in favor of Controls. 

I think at least in the Avatar example `withKnobs` can actually be removed (not sure why it was there), but that should solve that issue. Also I think the Knobs(number) will show the total number of knobs in that entire Component.stories.jsx file? Noticed it with `Button` where we only used 2 knobs in the `Loading` story but will show in the others.

Screenshot 2023-05-15 at 2 29 07 PM

Screenshot 2023-05-15 at 2 28 57 PM


It might take some time to update all the Stories to use the Controls arg structure and see the full benefits. Knobs will still work in the `Component.stories.jsx` files that still use them though.

File: .storybook/preview.js (hunk 6 above)
Comment (jasonbasuil) [reply .storybook/preview.js:8]: I would be happy to explore bringing Prettier into the DS. I may want to wait until this gets merged just to keep the diff manageable as it's getting pretty big. Perhaps we can get started on talking about what JS style normalizations we would eventually want to explore?

Comment (kyleshike) [reply .storybook/preview.js:8]: Yea I think that's a good call.  There's no need to it to be a part of this upgrade!

File: .nvmrc [hunk 14]
```diff
@@ -0,0 +1 @@
+16.20.0
```
[the remark this answers is not in the document]
Comment (jasonbasuil) [reply .nvmrc]: makes sense to me and enjoy the consistency

**Comment by jasonbasuil** [comment]
The most recent Chromatic build can be found here: https://62d040e741710e4f085e0647-yoxsmdmwaj.chromatic.com/?path=/docs/intro--docs

I combed through it one more time story by story to make sure nothing breaks on our docs or existing stories.

After this gets merged, will prepare a new release, designate `v2.0.0` to update `package.json` and then create the new release off `main`. It will then be ready to use on RS.

Follow ups:
- Continue transitioning `Component.stories.jsx` files to handle new Controls args. Knobs will continue to work as expected, but Controls won't be used fully until transitioned. I imagine after we deprecate all knobs we'll be able to remove `@storybook/addon-knobs` from our devDependencies.
- Look into testing out Prettier in the Design System (as a proof of concept) before introducing into RS.

Let me know if there's anything else that might need taking care of before moving this forward! @jeffbax @kyleshike @rroppolo

**Review by jeffbax — APPROVED** [review]
holy shit this is awesome

**Pull request merged**